tensorflow_datasets/core/image_visualizer.py

```diff
@@ -0,0 +1,222 @@
+      if label_key:
+        label = ex[label_key]
+        label_str = ds_info.features[label_key].int2str(label)
+        plt.xlabel("{} ({})".format(label_str, label))
+    plt.show()
+    return fig
+
+
+class ObjectVisulaizer(ImageVisualizer):
+  """Class for Object Detection Datasets"""
+
+  def _draw_text(self, ax, xy, txt, size=14):
+    """Draws text on the top-left corner of the bounding box"""
+    ax.text(*xy, txt, verticalalignment='top', \
+            color='white', fontsize=size, weight="normal", \
+            bbox=dict(facecolor='black', alpha=0.7, pad=0.3))
+
+  def _draw_rectangle(self, ax, bbox):
+    """Draws the bounding box"""
+    patches = lazy_imports_lib.lazy_imports.matplotlib.patches
+    ax.add_patch(patches.Rectangle(bbox[:2], \
+                *bbox[-2:], fill=False, \
+                edgecolor='red', lw=0.7))
+
+  def _bb_hw(self, bbox, height, width):
+    """converts bounding box to height and width"""
+    return [bbox[1]*width, bbox[0]*height, \
+            (bbox[3]*width)-(bbox[1]*width)+1, \
+            (bbox[2]*height)-(bbox[0]*height)+1]
+
+  def match(self, ds_info):
+    """Checks if dataset belongs to object detection category"""
+    image_keys = self._extract_keys(ds_info, features_lib.Image)
+
+    if not image_keys:
+      return False
+
+    sequence_keys = self._extract_keys(ds_info, features_lib.Sequence)
+
+    bbox_keys = []
+    for key in sequence_keys:
+      if isinstance(ds_info.features[key].feature, features_lib.FeaturesDict):
+        bbox_keys = self._extract_keys(ds_info, features_lib.BBoxFeature, key)
+
+    if bbox_keys:
+      return True
+
+    return False
+
+  def build(self, ds_info, ds, rows=3, cols=3,
+            plot_scale=3, image_key=None, bbox_label=True):
+    """Builds the plot for visualization"""
+
+    image_key = self._infer_image_key(ds_info, image_key)
+    sequence_keys = self._extract_keys(ds_info, features_lib.Sequence)
+
+    # Infer the BBox keys in the sequence feature connectors
+    bbox_keys = []
+    for key in sequence_keys:
+      if isinstance(ds_info.features[key].feature, features_lib.FeaturesDict):
+        bbox_keys = self._extract_keys(ds_info, features_lib.BBoxFeature, key)
+        if len(bbox_keys) > 0:
+          sequence_key = key
+          break
+
+    # Infer the label keys in the sequence feature connector (BBox Labels)
+    label_keys = self._extract_keys(ds_info, features_lib.ClassLabel,
+                                    sequence_key)
+
+    # Taking first label key since some datasets have multiple label keys
+    # for BBoxes as in voc dataset
+    label_key = label_keys[0] if len(label_keys) > 0 else None
+    if not label_key:
+      logging.info("Was not able to auto-infer label.")
+    bbox_key = bbox_keys[0] if len(bbox_keys) > 0 else None
+
+    num_examples = rows * cols
+    examples = list(dataset_utils.as_numpy(ds.take(num_examples)))
+
+    fig, plt = self._init_plot(rows, cols, plot_scale)
```

Comment:
Only the drawing part seems to be different between the two class, The loop seems to be the same, so this is what should have been factorised out too. Maybe something like:

```python
def _make_fig(draw_fn):
  ...
  for i, ex in enumerate(examples):
    ...
    draw_fn(fig, ...)
```